### PR TITLE
feat: use `cargo generate` instead of hardcoding examples source code

### PR DESCRIPTION
**Pull request by paulotten**
## Description of change

Use `cargo-generate` (as a library) to bootstrap copies of `shuttle-examples` apps.

Closes https://github.com/shuttle-hq/shuttle/issues/867.

## How Has This Been Tested (if applicable)?

Ran locally.

### CODE REVIEW

File: cargo-shuttle/src/init.rs
Comment (paulotten) [inline cargo-shuttle/src/init.rs:190]: I'll need to create an empty project for this in `shuttle-examples`

**Comment by paulotten** [comment]
@oddgrd what do you think of this approach using `cargo generate`?

It requires the user to have `cargo generate` installed, which I'll have to add documentation for.

In theory it's possible to include `cargo-generate` as a library instead. In practice I haven't been able to make the dependencies compile. (cargo-generate=0.18 requires git2=0.18, which conflicts with other things we use. cargo-generate=0.17 requires git2=0.16 which we use, but then fails to compile. cargo-generate=0.16 requires git2=0.14 which we is too low.)

**Comment by oddgrd** [comment]
The dependency conflict should also be easier to resolve after we remove the `cargo` dependency, which is in progress. The `cargo` dependency currently ties down the `git2` version.

**Comment by iulianbarbu** [comment]
@paulotten out of curiosity. The `cargo-shuttle` dependency on cargo was removed here: https://github.com/shuttle-hq/shuttle/pull/765/files. Would you still be blocked by the `cargo` dependency that we have in the other places, when trying to fix the conflicts coming from `git2` dependency?

**Comment by iulianbarbu** [comment]
Oh, that makes sense. Ok, thanks for the clarification! 🙏

**Comment by iulianbarbu** [comment]
#756  was resolved by #922 . @paulotten I am curious if we can continue with this PR. Do you have the time to take another stab at it? The idea is that we want to close #821 pretty soon and we're interested in getting this PR over the line. Let us know what you think

File: cargo-shuttle/src/init.rs
Comment (paulotten) [reply cargo-shuttle/src/init.rs:191]: Please see https://github.com/shuttle-hq/shuttle-examples/pull/57

Comment (jonaro00) [reply cargo-shuttle/src/init.rs:191]: I can fix a better (useful) version of https://github.com/shuttle-hq/shuttle-examples/pull/57 so that this is resolved.

Comment (jonaro00) [reply cargo-shuttle/src/init.rs:191]: https://github.com/shuttle-hq/shuttle-examples/pull/59

File: cargo-shuttle/src/lib.rs
Comment (iulianbarbu) [inline cargo-shuttle/src/lib.rs:220]:
```suggestion
        init::cargo_generate(
            path.clone(),
            project_args
                .name
                .as_ref()
                .expect("to have a project name provided"),
            framework,
        )?;
```

File: cargo-shuttle/src/init.rs
Comment (iulianbarbu) [inline cargo-shuttle/src/init.rs:195]:
```suggestion
pub fn cargo_generate(path: PathBuf, name: &ProjectName, framework: Template) -> Result<()> {
```

**Review by iulianbarbu — COMMENTED** [review]
For some reason it doesn't work well on my side. Doing an init for an `axum` project brings over the entire shuttle-examples repo. Can you reproduce it?

**Comment by jonaro00** [comment]
I also get the entire repo

File: cargo-shuttle/src/init.rs
Comment (paulotten) [reply cargo-shuttle/src/init.rs]: change applied manually

File: cargo-shuttle/src/lib.rs
Comment (paulotten) [reply cargo-shuttle/src/lib.rs]: change applied manually

**Comment by jonaro00** [comment]
You can bump the examples submodule to the latest on main :)

**Comment by paulotten** [comment]
> For some reason it doesn't work well on my side. Doing an init for an `axum` project brings over the entire shuttle-examples repo. Can you reproduce it?

Should be fixed now.

Now I'm noticing that the crate name isn't getting set correctly. Looking into it.

**Comment by paulotten** [comment]
> You can bump the examples submodule to the latest on main :)

Fixed, I think.

File: cargo-shuttle/src/lib.rs
Comment (jonaro00) [inline cargo-shuttle/src/lib.rs:221]: @paulotten This `cargo_shuttle_init` used to do the project name change. You need to restore that part of it.

**Comment by paulotten** [comment]
> Now I'm noticing that the crate name isn't getting set correctly. Looking into it.

Should be fixed. (Ended up using `toml-edit` to set it.)

**Review by jonaro00 — APPROVED** [review]
LGTM! Maybe just a tiny nitpick: the toml_edit imports could be at top of file.
I'm looking at if we can fix the failing rocket init tests and kill two birds with one stone.

**Comment by jonaro00** [comment]
> Looks like just a new line at end of file difference.

Yeah, but there are also some tests that fail sporadically. I think I've managed to reproduce it.

This can be merged if the team wants, and I'll add the fixes to the subsequent PR.

File: cargo-shuttle/Cargo.toml
Comment (chesedo) [inline cargo-shuttle/Cargo.toml:25]: Nice, to see this resolved!! :partying_face:

**Review by chesedo — COMMENTED** [review]
Always love to see a PR that removes code.

Small nit: maybe the `Shuttle.toml` should be deleted since it will override the project to something that already exists.

**Comment by paulotten** [comment]
> Ya, I can do that.

There should be an update to this PR for this commit https://github.com/shuttle-hq/shuttle/commit/65ab862983616adb17b9c59c4947ff2aee95bb7a

GitHub is having issues though. https://www.githubstatus.com/

> [Incident with Actions, Pull Requests and Webhooks](https://www.githubstatus.com/incidents/1g1gkh0qpyvs)

**Comment by paulotten** [comment]
> GitHub is having issues though.

Appears to be fixed. I see the expected commit now.

**Review by iulianbarbu — APPROVED** [review]
LGTM! Thanks a lot for this. 🙏

**Comment by jonaro00** [comment]
I saw now that two things are missing! The `init` flag to generate does not initialize a git repo like `cargo init` does. One more thing that cargo does is add a .gitignore with `target/`. We should probably add this aswell but we can also include `Secrets*.toml`!

**Comment by oddgrd** [comment]
> I saw now that two things are missing! The `init` flag to generate does not initialize a git repo like `cargo init` does. One more thing that cargo does is add a .gitignore with `target/`. We should probably add this aswell but we can also include `Secrets*.toml`!

Good catch! Should we include that here, or have you started on these in your new PR @jonaro00?

**Comment by jonaro00** [comment]
I have not started on those. You can do them here if you want.

**Comment by paulotten** [comment]
> One more thing that cargo does is add a .gitignore with `target/`.

`cargo_generate` has a `vcs` option. Setting it to `Git` however doesn't seem to create a `.gitignore`.

> We should probably add this aswell but we can also include `Secrets*.toml`!

`cargo_generate` isn't going to know to add that. I'll just write a `.gitignore` file manually.

**Comment by jonaro00** [comment]
Great.
The template could have a gitignore already. Should we append to the file if it exists? Or parse it and only add the lines if they are missing?

**Comment by paulotten** [comment]
> Good point! Another thought, can we use the git2 library for this?

It depends how complex we need to make this.

If we need something like this, I'd probably go reaching for a `gix` crate (Rust rewrite of git functionality). Maybe [gix_ignore](https://docs.rs/gix-ignore/latest/gix_ignore/#).

**Comment by jonaro00** [comment]
So shall we add a .gitignore with `target/` and `Secrets*.toml` to all examples? It seems reasonable that they have this as a standard now that cloning will be the default way of starting off.

**Comment by jonaro00** [comment]
> Setting it to Git however doesn't seem to create a .gitignore.

We can still pass that option so that a git repo is init'ed.

**Review by oddgrd — APPROVED** [review]
LGTM, thanks for finding the `cargo-generate` solution and for finishing this up so quickly after the cargo removal! :heart:

**Pull request merged**
